brian2genn/device.py

```diff
@@ -1649,6 +1649,38 @@ def process_state_monitors(self, directory, state_monitors, writer):
 
             self.state_monitor_models.append(sm)
 
+    def consolidate_pull_operations(self, run_regularly_operations):
+        def append(adict, name, step):
+            if name in adict:
+                adict[name].append(step)
+            else:
+                adict[name] = [step]
+        models_start = {}
+        models_end = {}
+        for sm in self.state_monitor_models:
+            if sm.when == 'start':
+                for varname in sm.variables:
+                    append(models_start, f'{varname}{sm.monitored}', sm.step)
+            else:
+                for varname in sm.variables:
+                    append(models_end, f'{varname}{sm.monitored}', sm.step)
+        for op in run_regularly_operations:
+            for varname in op['read']:
+                if varname not in ['t', 'dt']:
+                    owner_name = op['owner'].variables[varname].owner.name
+                    append(models_start, f'{varname}{owner_name}', op['step'])
+        # Shortcut: If a state is pulled on every turn, no need to list all steps
+        models_start = {key: [1] if 1 in val else list(set(val))
+                        for key, val in models_start.items()}
+        models_end = {key: [1] if 1 in val else list(set(val))
+                      for key, val in models_end.items()}
```

Comment:
Maybe replace `list(set(val))` by `ordered(set(val))`? I'm a bit worried that otherwise the order of the operations could change in subsequent runs, which would lead to unnecessary recompilation of the `engine.cpp` file.